src/waveform/waveform.h

```diff
@@ -98,6 +105,7 @@ class Waveform {
     }
     void setCompletion(int completion) {
         m_completion = completion;
+        m_saveState = Waveform::SaveState::Changing;
```

Comment:
This function is called from a different thread and will set m_completion atomically.On the other hand m_saveState is not protected against any race conditions. I also have doubts if it is a correct to manage the state of the waveform redundantly in different variables.